File: leetcode/28.java
class Solution {
    public int strStr(String haystack, String needle) {
        if (needle.length() == 0) return 0;
        Kmp kmp = new Kmp(haystack, needle);
        List<Integer> list = kmp.getMatchPoint();
        if (list.isEmpty()) return -1;
        else return list.get(0);
    }
    
    public class Kmp {

    private int[] s;
    private int[] p;
    private int[] t;
    private int[] a;

    public Kmp(String string, String pattern){
        int i,j;
        // pad arrays
        s = convertString(string, -1);
        p = convertString(pattern, -2);

        // construct function phi
        t = new int [p.length];
        for (i=0;i<p.length;++i) t[i] = -1;
        i=1; j=0;
        while (i < t.length){
            if (t[i] == -1) t[i] = j;
            if (j < 0) {++i; j=0;}
            else if (p[i] == p[j]) {++i; ++j;}
            else {j = t[j];}
        }
        // optimize t, the following loop can be commented out
       for (i=1;i<t.length;++i)
           if (t[i] >= 0 && p[i] == p[t[i]]) t[i] = t[t[i]];
           else t[i] = t[i];

        // matching, store result in a
        a = new int [s.length];
        for (i=0; i<a.length;++i) a[i] = -1;
        i = 0; j = 0;
        while (i < a.length){
            if (a[i] == -1) a[i]=j;
            if (j < 0) {++i; j=0;}
            else if (s[i] == p[j])  {++i; ++j;}
            else {j = t[j];}
        }
    }

    private int[] convertString(String string, int endSymbol){
        int[] result = new int[string.length() + 1];
        for (int i=0;i<string.length();++i) result[i] = string.charAt(i);
        result[string.length()] = endSymbol;
        return result;
    }

    public int[] getMatchResult(){
        return a;
    }

    public List<Integer> getMatchPoint(){
        List<Integer> result = new ArrayList<>();
        for (int i=0;i<a.length;++i)
            if (a[i] == p.length - 1) result.add(i-p.length+1);
        return result;
    }
}
}

File: leetcode/28.cpp
vector<int> kmp(string s, string p){
    // add -1 into p, -1 can not be in s
    p.push_back(-1);
    int n = s.length();
    int m = p.length();

    // construct function phi
    vector<int> phi(m);
    int i = 0, j = -1;
    phi[0] = -1;
    while (i < m-1){
        if (p[i+1] == p[j+1]) phi[++i] = ++j;
        else if (j >= 0) j = phi[j];
        else phi[++i] = -1;
    }

    // optimized fuction phi
    for (int i=0;i<m-1;++i)
        while (phi[i] != -1 && p[i+1] == p[phi[i]+1]) phi[i] = phi[phi[i]];

    // match p against s
    vector<int> f(n);
    i = -1; j = -1;
    while (i < n-1){
        if (s[i+1] == p[j+1]) f[++i] = ++j;
        else if (j >= 0) j = phi[j];
        else f[++i] = -1;
    }

    vector<int> ans;
    for (int i=0;i<n;++i)
        if (f[i] == m-2) ans.push_back(i - m + 2);
    return ans;
}


class Solution {
public:
    int strStr(string haystack, string needle) {
        vector<int> ans = kmp(haystack, needle);
        if (needle.empty()) return 0;
        if (ans.empty()) return -1;
        else return ans[0];
    }
};
